### PR TITLE
[V4] Query engine

### PR DESCRIPTION
**Pull request by alexandrebodin**
# Query Engine

This PR introduces a draft version of the Query Engine.

You can read it [here](https://github.com/strapi/rfcs/blob/v4/query-engine/rfcs/xxx-v4-query-engine.md)

### CODE REVIEW

File: rfcs/xxx-v4-query-engine.md [hunk 1]
```diff
@@ -0,0 +1,744 @@
+- Start Date: 2021-07-01
+- RFC PR: (leave this empty)
+
+# Summary
+
+In v4 we are build a new database layer to support the futur evolution of Strapi and we are introducing a new query engine with it.
```
Comment (gustavpursche) [inline rfcs/xxx-v4-query-engine.md:6]:
```suggestion
In v4 we are build a new database layer to support the future evolution of Strapi and we are introducing a new query engine along with it.
```

File: rfcs/xxx-v4-query-engine.md [hunk 2]
```diff
@@ -0,0 +1,744 @@
...
+
+The query engine should mostly be used by plugin developers, developers adding custom business logic to their applications and the Strapi core team.
+
+# Example
+
+```js
+qb.query("article").findMany({
```
Comment (gustavpursche) [inline rfcs/xxx-v4-query-engine.md:13]:
```suggestion
db.query("article").findMany({
```

File: rfcs/xxx-v4-query-engine.md [hunk 3]
```diff
@@ -0,0 +1,744 @@
...
+  },
+  populate: {
+    category: true,
+  },
+});
+```
+
+# Motivation
+
+We want to make sure we have a more flexible and powerfull database layer to build the futur features of Strapi.
```
Comment (gustavpursche) [inline rfcs/xxx-v4-query-engine.md:28]:
```suggestion
We want to make sure we have a more flexible and powerful database layer to build future features of Strapi onto.
```

File: rfcs/xxx-v4-query-engine.md [hunk 4]
```diff
@@ -0,0 +1,744 @@
...
+
+Moreover we want to provide a better experience and more possibilities for plugin developers and application developers to build advanced business logic into their plugins or applications.
+
+This layer is really made for database queries and is at a lower level than our previous version. This layer isn't abstracting a lot of things so we can have full control.
```
Comment (gustavpursche) [inline rfcs/xxx-v4-query-engine.md:32]:
```suggestion
This layer is really made for database queries and is at more low-level than our previous version. This layer isn't abstracting a lot of things so developers can have full control.
```

File: rfcs/xxx-v4-query-engine.md [hunk 5]
```diff
@@ -0,0 +1,744 @@
...
+
+# Detailed design
+
+## Query API
+
+To work with entities in the database you can user the Query api:
```
Comment (gustavpursche) [inline rfcs/xxx-v4-query-engine.md:38]:
```suggestion
To work with entities in the database you can use the Query api:
```

File: rfcs/xxx-v4-query-engine.md [hunk 6]
```diff
@@ -0,0 +1,744 @@
...
+```js
+const entries = await db.query("article").findMany({
+  select: ["title", "description"],
+  where: { title: "Hello World" },
+  orderBy: { title: "DESC" },
+  populate: { category: true },
+});
+```
+
+### `findWithCount(params)` => `[Entry[], number]`
```
Comment (gustavpursche) [inline rfcs/xxx-v4-query-engine.md:103]: Should this method be called `findManyWithCount` to be more consistent?

File: rfcs/xxx-v4-query-engine.md [hunk 7]
```diff
@@ -0,0 +1,744 @@
...
+### `delete(params)` => `Entry`
+
+Deletes one entry and returns it
+
+**Params**
+
+| Param    | Type                           | Description                                |
+| -------- | ------------------------------ | ------------------------------------------ |
+| select   | `string[]`                     | Select some attributes to return           |
+| populate | [`PopulateParam`](#populating) | Select which relations should be populated |
```
Comment (gustavpursche) [inline rfcs/xxx-v4-query-engine.md:184]: Does `populate` make sense for delete queries? I'm not sure what I'd expect from that ... does it control the `cascade` behavoir?

File: rfcs/xxx-v4-query-engine.md [hunk 8]
```diff
@@ -0,0 +1,744 @@
...
+// 12
+```
+
+### Avg,Min,Max,Sum
+
+> The rest of this API (avg,min,max,sum,groupby) will specified at a later time.
+
+## Query Builder
+
+When in need fore more control over the queries than with the Query API you can use the underlying query builder.
```
Comment (gustavpursche) [inline rfcs/xxx-v4-query-engine.md:307]:
```suggestion
When you are in need for more control over the queries the Query API generates, you can use the underlying query builder.
```

File: rfcs/xxx-v4-query-engine.md (hunk 7 above)
Comment (alexandrebodin) [reply rfcs/xxx-v4-query-engine.md:184]: This has the same behaviour create & update to be consistent

File: rfcs/xxx-v4-query-engine.md [hunk 9]
```diff
@@ -0,0 +1,744 @@
...
+    dynamiczoneA: true,
+    relation: someLogic || true,
+  },
+});
+```
+
+You can also apply where filters and select or populate nested relations
+
+```js
+db.query("article").findMany({
```
Comment (alexandrebodin) [inline rfcs/xxx-v4-query-engine.md:681]: NOTE: Extends the example a bit more

**Comment by brettwillis** [comment]
I like that population can be controlled for components etc, where currently only population of the top level model can be controlled.

Here are some initial comments:
1. I can see some filtering operators that not all database layers will support (e.g. `$endsWith`, `$contains`, and also complex combinations of and/or). Have you considered if there shall be some defined behaviour when a database connector doesn't support a query? Just throw an error?
2. Some database layers support pagination/continuation via continuation tokens. Tokens could be any type but typically a string. Have you considered supporting continuation tokens from previous queries, as an alternative to `offset`?


Some motivation behind point 2: I'm thinking of some database layers where use of `limit` can be costly, costing a "read operation" for every skipped entity. Pagination in the admin dashboard could be more effective with continuation tokens.

**Pull request closed without merging**